lib/patch-series.ts

```diff
@@ -231,31 +262,69 @@ export class PatchSeries {
         return mbox.split(separatorRegex);
     }
 
-    protected static removeDuplicateMimeVersionHeaders(mails: string[]): void {
+    protected static removeDuplicateHeaders(mails: string[]): void {
         mails.map((mail: string, i: number) => {
             const endOfHeader = mail.indexOf("\n\n");
             if (endOfHeader < 0) {
                 return;
             }
+
             let headers = mail.substr(0, endOfHeader + 1);
-            const needle = "\nMIME-Version: 1.0\n";
-            const offset = headers.indexOf(needle);
-            if (offset < 0) {
-                return;
-            }
-            let offset2 = headers.indexOf(needle, offset + 1);
-            if (offset2 < 0) {
-                return;
-            }
-            do {
-                headers = headers.substr(0, offset2 + 1)
-                    + headers.substr(offset2 + needle.length);
-                offset2 = headers.indexOf(needle, offset + 1);
-            } while (offset2 > 0);
+            singletonHeaders.forEach((header: ISingletonHeader) => {
+                headers = PatchSeries.stripDuplicateHeaders(headers, header);
+            });
+
             mails[i] = headers + mail.substr(endOfHeader + 1);
         });
     }
 
+    protected static stripDuplicateHeaders(headers: string,
+                                           header: ISingletonHeader): string {
+        const needle = "\n" + header.key + ":";
+        let offset;
+
+        if (headers.startsWith(`${header.key}:`)) {
+            offset = 0;
+        } else {
+            offset = headers.indexOf(needle) + 1;
+            if (!offset) {
+                return headers;
+            }
+        }
+
+        let endofKey = offset + needle.length - 1;
+        offset = headers.indexOf(needle, endofKey + 1);
+
+        if (offset < 0) {
+            return headers;
+        }
+
+        // extract values to determine if they match.
+        let endOfHdr = headers.indexOf("\n", endofKey);
+        const value1 = headers.substr(endofKey, endOfHdr - endofKey).trim();
+
+        do {
+            endofKey = offset + needle.length;
+            endOfHdr = headers.indexOf("\n", endofKey);
+            const value2 = headers.substr(endofKey,
+                                          endOfHdr - endofKey).trim();
+
+            if (value1 !== value2) {
+                if (0 >= header.values.indexOf(value2)) {
+                    console.log("Header values do not match for " +
+                        header.key + ": " + value1 +
+                        ":" + value2 + "\n" + headers);
+                }
+            }
+
+            // substr up to \n and concat from next \n
+            headers = headers.substr(0, offset) + headers.substr(endOfHdr);
```

Comment:
`offset` points at the newline just before the instance of `key` which we want to remove, so the first `substr()` grabs everything up to but not including that newline; `endOfHdr` points at the newline following the value of the instance of `key` we want to remove, so the second `substr()` grabs that newline and everything following it. Okay.